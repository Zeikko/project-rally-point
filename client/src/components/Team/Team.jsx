import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import * as propTypes from '../../constants/prop-types'
import Squad from '../Squad/Squad'

function Team(props) {
  const {
    players, team,
  } = props
  const playersOfTeam = players.filter(player => player.team === team)
  return (
    <div>
      <Heading>Team {team} ({playersOfTeam.length} players)</Heading>
      <Squad players={playersOfTeam} squad={1} />
      <Squad players={playersOfTeam} squad={2} />
      <Squad players={playersOfTeam} squad={3} />
    </div>
  )
}

Team.propTypes = {
  players: PropTypes.arrayOf(propTypes.player).isRequired,
  team: PropTypes.number.isRequired,
}

export default Team

const Heading = styled.div`
  font-size: 20px;
`
