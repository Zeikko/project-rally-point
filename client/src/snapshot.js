import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}