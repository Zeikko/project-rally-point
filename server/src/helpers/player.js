import db from '../db'

export async function getIsPlayerInGame(userId, gameId) {
  const player = await db('player')
    .select('id')
    .where({
      userId,
      gameId,
    })
  return player.length > 0
}

export async function pickPlayer(gameId, userId, team, squad, role) {
  return db('player')
    .update({
      team,
      role,
      squad,
    })
    .where({
      gameId,
      userId,
    })
    .returning('*')
}

export function getGamePlayers(gameId) {
  return db('player')
    .select([
      'user.id',
      'steamId',
      'displayName',
      'smallAvatarUrl',
      'country',
      'player.role',
      'player.team',
      'player.squad',
    ])
    .leftJoin('user', 'player.userId', 'user.id')
    .where({ gameId })
}