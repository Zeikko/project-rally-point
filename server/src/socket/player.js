import db from '../db'
import actions from '../../../common/actions.json'
import { startCaptainVote, getGameStatus, shouldStartCaptainVote, createNewGame } from '../helpers/game'
import logger from '../logger'
import gameStatuses from '../../../common/game-statuses'

export default function handlePlayer(io, socket, action) {
  switch (action.type) {
    case actions.GET_PLAYERS_REQUEST:
      return getPlayersRequest(io, socket, action.gameId)
    case actions.JOIN_GAME_REQUEST:
      return joinGameRequest(io, socket, action.gameId)
    case actions.LEAVE_GAME_REQUEST:
      return leaveGameRequest(io, socket, action.gameId)
    case actions.VOTE_CAPTAIN_REQUEST:
      return voteCaptainRequest(io, socket, action.gameId, action.playerId)
    default:
      return null
  }
}

export async function getPlayersRequest(io, socket, gameId) {
  try {
    const players = await getGamePlayers(gameId)
    socket.emit('action', { type: actions.GET_PLAYERS_SUCCESS, data: players })
  } catch (error) {
    logger.error(error)
    socket.emit('action', { type: actions.GET_PLAYERS_ERROR })
  }
}

async function joinGameRequest(io, socket, gameId) {
  try {
    const { status } = await getGameStatus(gameId)
    if (status !== gameStatuses.QUEUE) {
      throw Error('Game is already started')
    }
    await db('player').insert({
      gameId,
      userId: socket.userId,
    })
    socket.emit('action', { type: actions.JOIN_GAME_SUCCESS })
    await broadcastGamePlayersUpdate(io, action.gameId)
    const willStartCaptainVote = await shouldStartCaptainVote(gameId)
    if (willStartCaptainVote) {
      await createNewGame()
      const game = await startCaptainVote(gameId)
      io.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
    }
  } catch (error) {
    logger.error(error)
    socket.emit('action', { type: actions.JOIN_GAME_ERROR })
  }
}

async function leaveGameRequest(io, socket, gameId) {
  try {
    const game = await getGameStatus(gameId)
    if (game.status !== gameStatuses.QUEUE) {
      throw Error('Game is already started')
    }
    await db('player')
      .delete()
      .where({
        gameId,
        userId: socket.userId,
      })
    socket.emit('action', { type: actions.LEAVE_GAME_SUCCESS })
    await broadcastGamePlayersUpdate(io, gameId)
  } catch (error) {
    logger.error(error)
    socket.emit('action', { type: actions.LEAVE_GAME_ERROR })
  }
}

async function voteCaptainRequest(io, socket, gameId, playerId) {
  try {
    console.log(gameId)
    const { status } = await getGameStatus(gameId)
    if (status !== gameStatuses.VOTE_CAPTAINS) {
      throw Error('Game is not in captain vote')
    }
    console.log(gameId, socket.userId, playerId)
    await db('captainVote').insert({
      gameId,
      voterId: socket.userId,
      votedId: playerId
    })
    socket.emit('action', { type: actions.VOTE_CAPTAIN_SUCCESS })
    /*await broadcastGamePlayersUpdate(io, action.gameId)
    const willStartCaptainVote = await shouldStartCaptainVote(gameId)
    if (willStartCaptainVote) {
      await createNewGame()
      const game = await startCaptainVote(gameId)
      io.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
    }*/
  } catch (error) {
    console.log(error)
    logger.error(error)
    socket.emit('action', { type: actions.VOTE_CAPTAIN_ERROR })
  }
}

function getGamePlayers(gameId) {
  return db('player')
    .select('user.*')
    .leftJoin('user', 'player.userId', 'user.id')
    .where({ gameId })
}

async function broadcastGamePlayersUpdate(io, gameId) {
  const players = await getGamePlayers(gameId)
  io.emit('action', { type: actions.GET_PLAYERS_SUCCESS, data: players })
}
