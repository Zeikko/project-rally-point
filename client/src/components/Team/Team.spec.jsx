import React from 'react'
import { mount } from 'enzyme'
import { matchComponentToSnapshot } from '../../test/snapshot'
import Team from './Team'
import player from '../../../../fixtures/player.json'

describe('Team', () => {
  it('renders correctly', () => {
    matchComponentToSnapshot(<Team team={1} players={[{ ...player, team: 1 }]} />)
  })
})
