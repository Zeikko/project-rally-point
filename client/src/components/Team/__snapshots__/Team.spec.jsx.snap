// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team renders correctly 1`] = `
.c1 {
  font-size: 20px;
}

.c0 {
  font-size: 20px;
}

<div>
  <div
    className="c0"
  >
    Team 
    1
     (
    1
     players)
  </div>
  <div>
    <div
      className="c1"
    >
      Squad 
      1
       (
      0
       players)
    </div>
  </div>
  <div>
    <div
      className="c1"
    >
      Squad 
      2
       (
      0
       players)
    </div>
  </div>
  <div>
    <div
      className="c1"
    >
      Squad 
      3
       (
      0
       players)
    </div>
  </div>
</div>
`;
